matrix-api-r0.3.0.el

```diff
@@ -0,0 +1,480 @@
+                                  &key (visibility "private") alias name topic invite preset (is-direct t))
+  "Create new room on SESSION.
+When IS-DIRECT is non-nil, set that flag on the new room."
+  ;; https://matrix.org/docs/spec/client_server/r0.3.0.html#id190
+
+  ;; MAYBE: Add other parameters: invite_3pid, creation_content,
+  ;; initial_state.  Not sure how useful these would be for us.
+
+  (matrix-post session 'createRoom (a-list 'visibility visibility
+                                           'room_alias_name alias
+                                           'name name
+                                           'topic topic
+                                           'preset preset
+                                           'is-direct is-direct)
+               #'matrix-create-room-callback))
+
+(matrix-defcallback create-room matrix-session
+  "Callback for create-room.
+Add new room to SESSION."
+  ;; TODO: Should we add the room directly here, or should we do that after /sync?
+  :slots (rooms)
+  :body (pcase-let* (((map room_id) data)
+                     (room (matrix-room :session session
+                                        :id room_id)))
+          (push room rooms)))
+
+(cl-defmethod matrix-send-message ((room matrix-room) message)
+  "Send MESSAGE to ROOM."
+  ;; https://matrix.org/docs/spec/client_server/r0.3.0.html#id182
+  (with-slots* (((id session) room)
+                ((txn-id) session))
+    ;; Use `with-slots*' instead of `pcase-let*' so we can `incf' the txn-id.
+    (let* ((type "m.room.message")
+           (content (a-list 'msgtype "m.text"
+                            'body message))
+           (txn-id (cl-incf txn-id))
+           (endpoint (format "rooms/%s/send/%s/%s"
+                             id type txn-id)))
+      (matrix-put session endpoint content
+                  (apply-partially #'matrix-send-message-callback room)))))
+
+(matrix-defcallback send-message matrix-room
+  "Callback for send-message."
+  ;; For now, just log it, because we'll get it back when we sync anyway.
+  :slots (id)
+  :body (matrix-log "Message \"%s\" sent to room %s. Event ID: %s"
+                    id message (a-get data 'event_id)))
+
+(cl-defmethod matrix-leave ((room matrix-room))
+  "Leave room."
+  ;; https://matrix.org/docs/spec/client_server/r0.3.0.html#id203
+  (with-slots (id session) room
+    (let* ((endpoint (format "rooms/%s/leave" id)))
+      (matrix-post session endpoint nil
+                   (apply-partially #'matrix-leave-callback room)))))
+
+(matrix-defcallback leave matrix-room
+  "Leave room callback."
+  :slots (session)
+  ;; TODO: Verify that this works in more circumstances.  `equal' is
+  ;; used, and it works in the test, but will it always work?
+  :body (object-remove-from-list session :rooms room))
+
+(cl-defmethod matrix-forget ((room matrix-room))
+  "Forget ROOM."
+  ;; TODO: Maybe use room ID instead of object, since if we've left a
```

Comment:
FWIW, I don't think that forgetting rooms is a feature that we have to target, at least at first, so I don't think this matters. It is nice to have though (and useful for the test bot)

Even the android client can't forget rooms as of now :)

Reply:
Yeah, I don't think exposing this in the client is a high priority, but it helps for testing.  However, I think that I need to make the testing a bit more robust anyway, because it needs to be able to test existing room history.  So I think I need to set up a second test account that will remain in a room with history...  Kind of reluctant to do that, but I think it will solve the problem.  What I'd really like to do is for the test to register an account, test with it, and then deactivate the account, but since they use CAPTCHAs, I guess that's out.